### PR TITLE
Better bounds checks in ctcloss

### PR DESCRIPTION
**Pull request by ashermancinelli**
Adds better bounds checks for target lengths in CTC loss, checks for integral types for target and prediction lengths, and adds tests for each, according to #15946

### CODE REVIEW

**Comment by ashermancinelli** [comment]
cc @t-vi

File: aten/src/ATen/native/LossCTC.cpp [hunk 1]
```diff
@@ -163,6 +163,7 @@ std::tuple<Tensor, Tensor> ctc_loss_cpu_template(const Tensor& log_probs, const
 template<typename scalar_t, ScalarType target_scalar_type>
 Tensor ctc_loss_backward_cpu_template(const Tensor& grad_out, const Tensor& log_probs, const Tensor& targets, IntList input_lengths, IntList target_lengths,
                                       const Tensor& neg_log_likelihood, const Tensor& log_alpha, int64_t BLANK) {
+
```
Comment (t-vi) [inline aten/src/ATen/native/LossCTC.cpp]: Nit: I think this might go...

File: aten/src/ATen/native/LossCTC.cpp [hunk 2]
```diff
@@ -364,6 +365,10 @@ Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets, IntList input_le
 
 // Convenience function accepting Tensors
 Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets, const Tensor& input_lengths, const Tensor& target_lengths, int64_t BLANK, int64_t reduction) {
+
```
Comment (t-vi) [inline aten/src/ATen/native/LossCTC.cpp]: This, too. (It's probably from me, but I think starting functions with a blank line doesn't match the usual coding style.)

File: test/test_nn.py [hunk 3]
```diff
@@ -4237,6 +4237,10 @@ def test_CTCLoss_cudnn(self):
             res2 = torch.nn.functional.ctc_loss(log_probs, targets.cuda().long(), input_lengths, target_lengths)
         self.assertEqual(res, expected)
         self.assertEqual(res2, res)
+        with self.assertRaises(RuntimeError):
+            input_lengths = torch.tensor(input_lengths).type('torch.FloatTensor')
```
Comment (t-vi) [inline test/test_nn.py]: - I think we might move these to their own `test_...` method that isn't `cudnn`.
- Can you use `.float()` or `.to(dtype=torch.float)` instead, please? I think `.type` is a bit legacyish.
- Also I'd probably do the two float-tensor things separately.
- We would also want a test for the maximum size checks. Could you please add that?
  For those, we do want to check cpu and gpu, you could look in the other tests for a pattern for how to check both.

**Review by t-vi — COMMENTED** [review]
Thanks for sending this!
The code changes themselves look good to me, I think it would be good to add a few more tests.

File: test/test_nn.py (hunk 3 above)
Comment (ashermancinelli) [reply test/test_nn.py]: Working on those right now

**Review by ezyang — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

File: test/test_nn.py [hunk 4]
```diff
@@ -4238,6 +4238,35 @@ def test_CTCLoss_cudnn(self):
...
+        with self.assertRaises(RuntimeError):
+            target_lengths = target_lengths.to(dtype=torch.float)
+            torch.nn.functional.ctc_loss(log_probs, targets, input_lengths, target_lengths)
+
+    @unittest.skipIf(not TEST_CUDA, 'CUDA not available')
+    def test_CTCLoss_lengthchecks_cuda(self):
+        target_lengths = [30, 25, 20]
+        input_lengths = [50, 50, 50]
+        targets = torch.randint(1, 15, (3, 29), dtype=torch.int).to(dtype=torch.cuda.FloatTensor)
+        log_probs = torch.randn(50, 3, 15, dtype=torch.float).log_softmax(2).to(dtype=torch.cuda.FloatTensor)
```
Comment (t-vi) [inline test/test_nn.py]: You'd want `.to(dtype=torch.float, device='cuda')`, or really
`targets = torch.randint(1, 15, (3, 29), dtype=torch.long, device='cuda')` (targets should be on cuda but long to hit the native version).
Similarly, you'd just add `device=cuda` to the randn and drop the `.to(...)`.

**Comment by ashermancinelli** [comment]
I dont know why `TestSparseNormalize.test_sparse_normalize` is the only test that failed... Pulling upstream to see if that fixes it

**Comment by t-vi** [comment]
I wouldn't worry too much about the failure for https://github.com/pytorch/pytorch/pull/16269/commits/31fdb048c25433cc76126fd5e1481fbabde2676c - the rocm CI builds seem to be more prone to un-related build errors (and unlike the circle ci, I'm not sure they automatically retry).

**Comment by t-vi** [comment]
So now it all seems good for me, the build failure is indeed unrelated. Thanks @ashermancinelli !

**Comment by ashermancinelli** [comment]
@t-vi Yep! Let me know if there's anything else I need to do to get it merged.
